pandas/core/groupby/groupby.py

```diff
@@ -4990,25 +4997,67 @@ def shift(
         else:
             axis = 0
 
-        if freq is not None or axis != 0:
-            f = lambda x: x.shift(periods, freq, axis, fill_value)
-            return self._python_apply_general(f, self._selected_obj, is_transform=True)
+        if is_list_like(periods):
+            periods = cast(list, periods)
+            if axis == 1:
+                raise ValueError(
+                    "If `periods` contains multiple shifts, `axis` cannot be 1."
+                )
+            if len(periods) == 0:
+                raise ValueError("If `periods` is an iterable, it cannot be empty.")
+            from pandas.core.reshape.concat import concat
 
-        if fill_value is lib.no_default:
-            fill_value = None
-        ids, _, ngroups = self.grouper.group_info
-        res_indexer = np.zeros(len(ids), dtype=np.int64)
+            add_suffix = True
+        else:
+            if not isinstance(periods, int):
+                raise TypeError(
+                    f"Periods must be integer, but {periods} is {type(periods)}."
+                )
+            periods = [periods]
+            add_suffix = False
+
+        shifted_dataframes = []
+        for period in periods:
+            if not isinstance(period, int):
+                raise TypeError(
+                    f"Periods must be integer, but {period} is {type(period)}."
+                )
+            if freq is not None or axis != 0:
+                f = lambda x: x.shift(
+                    period, freq, axis, fill_value  # pylint: disable=cell-var-from-loop
```

Comment:
Can you address this pylint instead of ignoring?

Reply:
Cant you just define this before the for loop?

Reply:
Sorry, I don't get what you mean. Define what?

Reply:
This warnings mean that `f` is defined each time unnecessarily in a loop. That means it could be just defined once outside the loop

Reply:
Sorry, I still don't get what you mean. 

The warning refers to `period` being defined via the loop. If we defined the lambda before the loop, we have to feed it `period` still somehow at every iteration, so we would have to for example `partial` it right here so we can then pass it to `_python_apply_general` next line. Right?

Lots of extra code instead of telling `pylint` not to worry about `period` because it's used right after.

Reply:
Maybe I'm being a bit dim here.

Reply:
👍